phd-tests/tests/src/stats.rs

```diff
@@ -0,0 +1,495 @@
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::sync::Mutex;
+use std::time::Duration;
+
+use phd_testcase::*;
+use tracing::trace;
+use uuid::Uuid;
+
+use chrono::{DateTime, Utc};
+use dropshot::endpoint;
+use dropshot::ApiDescription;
+use dropshot::ConfigDropshot;
+use dropshot::HttpError;
+use dropshot::HttpResponseCreated;
+use dropshot::HttpServer;
+use dropshot::HttpServerStarter;
+use dropshot::RequestContext;
+use dropshot::TypedBody;
+use omicron_common::api::internal::nexus::ProducerEndpoint;
+use omicron_common::api::internal::nexus::ProducerKind;
+use omicron_common::api::internal::nexus::ProducerRegistrationResponse;
+use oximeter::types::{ProducerResults, ProducerResultsItem, Sample};
+use oximeter::{Datum, FieldValue};
+use slog::Drain;
+use slog::Logger;
+
+fn test_logger() -> Logger {
+    let dec = slog_term::PlainSyncDecorator::new(slog_term::TestStdoutWriter);
+    let drain = slog_term::FullFormat::new(dec).build().fuse();
+    let log = Logger::root(drain, slog::o!("component" => "fake-cleanup-task"));
+    log
+}
+
+// Re-registration interval for tests. A long value here helps avoid log spew
+// from Oximeter, which will re-register after about 1/6th of this interval
+// elapses.
+const INTERVAL: Duration = Duration::from_secs(300);
+
+// For convenience when comparing times below.
+const NANOS_PER_SEC: f64 = 1_000_000_000.0;
+
+struct PropolisOximeterSampler {
+    addr: std::net::SocketAddr,
+    uuid: Uuid,
+}
+
+struct FakeNexusContext {
+    sampler: Arc<Mutex<Option<PropolisOximeterSampler>>>,
+}
+
+#[derive(Copy, Clone, Debug, Eq, Hash, PartialEq)]
+enum VcpuState {
+    Emulation,
+    Run,
+    Idle,
+    Waiting,
+}
+
+impl VcpuState {
+    fn from_oximeter_state_name(name: &str) -> Self {
+        match name {
+            "emulation" => VcpuState::Emulation,
+            "run" => VcpuState::Run,
+            "idle" => VcpuState::Idle,
+            "waiting" => VcpuState::Waiting,
+            other => {
+                panic!("unknown Oximeter vpcu state name: {}", other);
+            }
+        }
+    }
+}
+
+#[derive(Default)]
+struct VcpuUsageMetric {
+    metrics: HashMap<VcpuState, u64>,
+}
+
+/// A collection of the stats produced by `propolis-server`'s Oximeter producer.
+///
+/// Oximeter producers produce a series of lists of samples, where each list
+/// of samples is conceptually distinct but may still be interesting to
+/// tset. In `propolis-server`, the first list of samples will be
+/// `virtual_machine:vcpu_usage`, which may be blank of kstats have not been
+/// sampled since the last producer poll. The second list of samples
+/// will be `virtual_machine:reset`.
+///
+/// `VirtualMachineMetrics` collects these all back together into a single view
+/// to test against. See [`VirtualMachineMetrics::add_producer_result`] as the
+/// means to accumulate samples into this struct.
+struct VirtualMachineMetrics {
```

Comment:
i'm right on the fence for this being buried away in `stats.rs` or more formally part of PHD. probably depends on how much more we expect propolis-sever to provide?

Reply:
With the caveat that I haven't looked at as many Oximeter interfaces or internals as I probably should have, I think this split looks pretty reasonable. Someday we may want to see about streamlining this a bit (so that, if I do all the work to define a new metric series in Propolis, it just "lights up" in PHD without my having to do anything), but I think this is a really good place to start. Thanks for reorganizing these bits!